src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -0,0 +1,253 @@
+import org.apache.cassandra.utils.FBUtilities;
+
+public class QueryController
+{
+    private final ColumnFamilyStore cfs;
+    private final ReadCommand command;
+    private final QueryContext queryContext;
+    private final TableQueryMetrics tableQueryMetrics;
+    private final RowFilter filterOperation;
+    private final List<DataRange> ranges;
+    private final AbstractBounds<PartitionPosition> mergeRange;
+
+    public QueryController(ColumnFamilyStore cfs,
+                           ReadCommand command,
+                           RowFilter filterOperation,
+                           QueryContext queryContext,
+                           TableQueryMetrics tableQueryMetrics)
+    {
+        this.cfs = cfs;
+        this.command = command;
+        this.queryContext = queryContext;
+        this.tableQueryMetrics = tableQueryMetrics;
+        this.filterOperation = filterOperation;
+        this.ranges = dataRanges(command);
+        DataRange first = ranges.get(0);
+        DataRange last = ranges.get(ranges.size() - 1);
+        this.mergeRange = ranges.size() == 1 ? first.keyRange() : first.keyRange().withNewRight(last.keyRange().right);
+    }
+
+    public TableMetadata metadata()
+    {
+        return command.metadata();
+    }
+
+    RowFilter filterOperation()
+    {
+        return this.filterOperation;
+    }
+
+    /**
+     * @return token ranges used in the read command
+     */
+    List<DataRange> dataRanges()
+    {
+        return ranges;
+    }
+
+    /**
+     * Note: merged range may contain subrange that no longer belongs to the local node after range movement.
+     * It should only be used as an optimization to reduce search space. Use {@link #dataRanges()} instead to filter data.
+     *
+     * @return merged token range
+     */
+    AbstractBounds<PartitionPosition> mergeRange()
+    {
+        return mergeRange;
+    }
+
+    /**
+     * @return indexed {@code IndexContext} if index is found; otherwise return non-indexed {@code IndexContext}.
+     */
+    public IndexContext getContext(RowFilter.Expression expression)
+    {
+        StorageAttachedIndex index = getBestIndexFor(expression);
+
+        if (index != null)
+            return index.getIndexContext();
+
+        return new IndexContext(cfs.metadata().keyspace,
+                                cfs.metadata().name,
+                                cfs.metadata().partitionKeyType,
+                                cfs.metadata().comparator,
+                                expression.column(),
+                                IndexTarget.Type.VALUES,
+                                null);
+    }
+
+    public StorageAttachedIndex getBestIndexFor(RowFilter.Expression expression)
+    {
+        return cfs.indexManager.getBestIndexFor(expression, StorageAttachedIndex.class).orElse(null);
+    }
+
+    public UnfilteredRowIterator getPartition(PrimaryKey key, ReadExecutionController executionController)
```

Comment:
```suggestion
    public UnfilteredRowIterator getRow(PrimaryKey key, ReadExecutionController executionController)
```
(...assuming we don't batch row reads in the same SPRC)

Reply:
A lot of these methods seem overly specific. I'm moving towards `queryStorage` for this one as it would cover cases where we read either a full partition or a row.